internal/server/resource/resource.go

```diff
@@ -65,42 +71,49 @@ type RankingInfo struct {
 
 // Update search index, given a stat var (group) node ID and string.
 func (index *SearchIndex) Update(
-	nodeID string, nodeString string, displayName string, isSvg bool) {
+	nodeID string, nodeString string, displayName string, isSvg bool, synonymMap map[string][]string) {
 	processedTokenString := strings.ToLower(nodeString)
 	processedTokenString = strings.ReplaceAll(processedTokenString, ",", " ")
 	tokenList := strings.Fields(processedTokenString)
-	addtionalTokens := []string{}
+	// Create a set of tokens from the tokens in tokenList and their synonyms
+	tokens := map[string]struct{}{}
 	for _, token := range tokenList {
-		if strings.HasSuffix(token, "s") {
-			addtionalTokens = append(addtionalTokens, strings.TrimSuffix(token, "s"))
-		}
-		if strings.HasSuffix(token, "es") {
-			addtionalTokens = append(addtionalTokens, strings.TrimSuffix(token, "es"))
-		}
-		if strings.HasSuffix(token, "ies") {
-			addtionalTokens = append(addtionalTokens, strings.TrimSuffix(token, "ies"))
+		tokens[token] = struct{}{}
+		if synonymList, ok := synonymMap[token]; ok {
+			for _, synonym := range synonymList {
+				tokens[synonym] = struct{}{}
+			}
 		}
 	}
-	tokenList = append(tokenList, addtionalTokens...)
 	approxNumPv := len(strings.Split(nodeID, "_"))
 	if approxNumPv == 1 {
 		// when approxNumPv is 1, most likely a non human curated PV
 		approxNumPv = nonHumanCuratedNumPv
 	}
 	// Ranking info is only dependent on a stat var (group).
 	index.Ranking[nodeID] = &RankingInfo{approxNumPv, displayName}
-	// Populate token to stat var map.
-	for _, token := range tokenList {
+	// Populate trie with each token
+	for token := range tokens {
+		currNode := index.RootTrieNode
+		for _, c := range token {
+			if currNode.ChildrenNodes == nil {
+				currNode.ChildrenNodes = map[rune]*TrieNode{}
+			}
+			if _, ok := currNode.ChildrenNodes[c]; !ok {
+				currNode.ChildrenNodes[c] = &TrieNode{}
+			}
+			currNode = currNode.ChildrenNodes[c]
+		}
 		if isSvg {
```

Comment:
this should block be inside the for loop above? Now only the last currNode has SvgIds

Reply:
My current implementation is for only the last currNode to have the SvgIds. I thought about adding the SvgId to every node along the way but I think that will use more space?

Reply:
My current implementation is for only the last currNode to have the SvgIds. I thought about adding the SvgId to every node along the way but I think that will use more space?

Reply:
For example when user search for "wom" (for "women"), would it have linked SvgIds?

Reply:
yes, because when user searches for "wom", the function will get to the "m: node and then traverse the rest of the subtrie that is rooted at the "m" node and get all the SvgIds and SvIds

Reply:
I see, that makes sense.